src/frontend/context.ts

```diff
@@ -0,0 +1,6 @@
+import { createContext } from "react";
+import type { MutatorDefs, Replicache } from "replicache";
+
+export const ReplicacheContext = createContext<Replicache<MutatorDefs> | null>(
```

Comment:
Maybe someone with a bit more Typescript experience can help out here. I'm not sure how to define this context such that the mutator types can be set dynamically ahead of time. Basically, to use this and get the correct editor autocomplete/hints I have to write a variation the following every time I grab the context:

```javascript
const replicache = useContext(ReplicacheContext) as Replicache<MyMutationsType>;
```

Reply:
Without it I see this:

<img width="982" alt="Screen Shot 2022-08-09 at 20 02 58" src="https://user-images.githubusercontent.com/2367158/183787845-2f1d1db2-eb12-46ad-b997-f94143a83471.png">

Reply:
`MutatorDefs` is just a dictionary type. It doesn't specify any particular mutators. `ReplicacheContext` needs to be generic over the specific `M extends MutatorDefs` for the users' mutators. So it can't be constant like this. Not sure exactly how to connect that to React's API though.

Reply:
We could use a Context creator function to pass along the generic. Something like this:

```javascript
function createReplicacheContext<M extends MutatorDefs>() {
  const ReplicacheContext = createContext<Replicache<M> | null>(null);
  
  function ReplicacheProvider({ children, ...options }) {
    const replicache = useReplicache<M>(options);
    return <ReplicacheContext.Provider>{children}</ReplicacheContext.Provider>
  }

  return { ReplicacheContext, ReplicacheProvider };
}
```

But the DX for that isn't great, and it doesn't solve the problem I mentioned above.

Reply:
I spent a bunch of time thinking about this.

It seems like in order for type safety to work correctly, the creation of `ReplicacheContext` is going to have to lie in the app, not in `replicache-nextjs` because (a) it has to be constant, (b) you want a different one for each `M`. That is, it's not really a `ReplicacheContext`, it's a `ReplicacheContext<M>` for some M.

You could then pass this context into `ReplicacheProvider` but at that point, I'm not sure what the value is. You can just as easily say, in your app:

```ts
const ReplicacheContext = createContext<Replicache<M> | null>(null);
```

and then later:

```ts
const rep = useReplicache({name, options});

<ReplicacheContext.Provider value={rep}>
  {children}
</ReplicacheContext>
```

Am I missing something?

You're only saving one line of code by trying to abstract it into replicache-nextjs and I'd argue the abstraction is not making it easier to follow.

Reply:
Like I get the dx win if there was a way to avoid that declaration of `ReplicacheContext` in your app, but I'm not seeing how to do that while retaining typesafety.

Reply:
I came to the same conclusion. Feel free to close this.